HeterogeneousCore/SonicCore/src/SonicClientBase.cc

```diff
@@ -0,0 +1,66 @@
+#include "HeterogeneousCore/SonicCore/interface/SonicClientBase.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+SonicClientBase::SonicClientBase(const edm::ParameterSet& params)
+    : allowedTries_(params.getUntrackedParameter<unsigned>("allowedTries", 0)) {
+  std::string modeName(params.getParameter<std::string>("mode"));
+  if (modeName == "Sync")
+    mode_ = SonicMode::Sync;
+  else if (modeName == "Async")
+    mode_ = SonicMode::Async;
+  else if (modeName == "PseudoAsync")
+    mode_ = SonicMode::PseudoAsync;
+  else
+    throw cms::Exception("Configuration") << "Unknown mode for SonicClient: " << modeName;
+
+  //get correct dispatcher for mode
+  if (mode_ == SonicMode::Sync or mode_ == SonicMode::Async)
+    dispatcher_ = std::make_unique<SonicDispatcher>(this);
+  else if (mode_ == SonicMode::PseudoAsync)
+    dispatcher_ = std::make_unique<SonicDispatcherPseudoAsync>(this);
+}
+
+void SonicClientBase::setDebugName(const std::string& debugName) {
+  debugName_ = debugName;
+  fullDebugName_ = debugName_;
+  if (!clientName_.empty())
+    fullDebugName_ += ":" + clientName_;
+}
+
+void SonicClientBase::start(edm::WaitingTaskWithArenaHolder holder) {
+  holder_ = std::move(holder);
+  tries_ = 0;
+  if (!debugName_.empty())
+    t0_ = std::chrono::high_resolution_clock::now();
+}
+
+void SonicClientBase::finish(bool success, std::exception_ptr eptr) {
+  //retries are only allowed if no exception was raised
+  if (!success and !eptr) {
+    ++tries_;
+    //if max retries has not been exceeded, call evaluate again
+    if (tries_ < allowedTries_) {
+      evaluate();
+      //avoid calling doneWaiting() twice
+      return;
+    }
+    //prepare an exception if exceeded
+    else {
+      cms::Exception ex("SonicCallFailed");
+      ex << "call failed after max " << tries_ << " tries";
+      eptr = make_exception_ptr(ex);
+    }
+  }
+  if (!debugName_.empty()) {
+    auto t1 = std::chrono::high_resolution_clock::now();
+    edm::LogInfo(fullDebugName_) << "Client time: "
+                                 << std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0_).count();
+  }
+  holder_.doneWaiting(eptr);
+}
+
+void SonicClientBase::fillBasePSetDescription(edm::ParameterSetDescription& desc, bool allowRetry) {
+  desc.add<std::string>("mode");
+  if (allowRetry)
+    desc.addUntracked<unsigned>("allowedTries", 0);
```

Comment:
is there a difference between 0 or 1 tries ?

Reply:
No.

Reply:
Thanks, good to know.
Pity for the indentation that `clang-format` imposes :-(

Reply:
Mhm, sorry, this was meant for the `ifValue` above :-)

Reply:
Indeed, the formatting is rather ugly. Maybe the core team can come up with a simpler syntax in the case where only `EmptyGroupDescription` is used.

Reply:
> Maybe the core team can come up with a simpler syntax in the case where only `EmptyGroupDescription` is used.

Please open an issue.

Reply:
I was waiting for you to write that :-)

Reply:
See #30932

Reply:
I know I sound like a broken record :)